src/geometry.rs

```
@@ -68,17 +142,53 @@ impl Vector {
     }
 }
 
-// impl Triangle {
-//     fn Intersects(&self, ray: &Ray) -> bool {}
-//     fn Intersects(&self, ray: &Ray) -> (bool, Point) {}
-//     ???
-//
-//     fn Area() -> f64
-//
-//     fn Normle() -> Vector
-//
-//     fn BaracentricCoordinates() -> ???
-// }
+impl Triangle {
+    fn area(&self) -> f64 {
+        let side_one = self.0 - &self.1;
+        let side_two = self.0 - &self.2;
+        (side_one.vector_product(&side_two)).norm() / 2.0
+    }
+
+    fn baracentric_coordinates(&self, point: &Point) -> (f64, f64, f64) {
+        let basic_normale = self.normale(&point).unwrap();
+        let normale_0 = (*point - &self.1).vector_product(&(*point - &self.2));
+        let area_0 = normale_0.norm()
+            * (if normale_0.scalar_product(&basic_normale) < 0.0 {
+                -1.
+            } else {
+                1.
+            });
+        let normale_1 = (*point - &self.2).vector_product(&(*point - &self.0));
+        let area_1 = normale_1.norm()
+            * (if normale_1.scalar_product(&basic_normale) < 0.0 {
+                -1.
+            } else {
+                1.
+            });
+        let normale_2 = (*point - &self.0).vector_product(&(*point - &self.1));
+        let area_2 = normale_2.norm()
+            * (if normale_2.scalar_product(&basic_normale) < 0.0 {
+                -1.
+            } else {
+                1.
+            });
+        let sum = area_0 + area_1 + area_2;
+        (area_0 / sum, area_1 / sum, area_2 / sum)
+    }
+}
+
+impl GeometricObject for Triangle {
+    fn intersects(&self, ray: &Ray) -> Option<Point> {
+        // I have no definition of Ray right now
+        panic!("Not implemented!");
+    }
+
+    fn normale(&self, point: &Point) -> Option<Vector> {
+        let side_one = self.0 - &self.1;
+        let side_two = self.0 - &self.2;
+        Some(side_one.vector_product(&side_two).normalize())
```

Why not to panic instead of returning none?

---

I think the proper way would be to use `Result`. Panic is the worst choice possible. If we can tell that there is something wrong in the other way, then we should use it.